README.rst
==========================================================
WSGI Middleware to Support X-Accel-Redirect |build status|
==========================================================

Description
===========

Nginx_ has a mechanism whereby, by returning an `X-Accel-Redirect`_ header,
a web app may convince nginx to serve a static file directly.
`PEP 333`_ defines a mechanism whereby an app server (or middleware)
can offer to serve files directly by providing ``wsgi.file_wrapper``
in the ``environ``.
This package provides a piece of WSGI middleware to take advantage of those
two mechanisms.

Despite the name of the package, currently this middleware only works
with nginx’s ``X-Accel-Redirect`` mechanism. (It would probably be
straightforward to generalize it so that it works ``X-Sendfile``, but
at the moment, I have no need for that.)

.. _Nginx: http://nginx.org/en/
.. _X-Accel-Redirect: http://wiki.nginx.org/X-accel
.. _PEP 333: http://www.python.org/dev/peps/pep-0333/
.. _wsgi.file_wrapper:
     http://www.python.org/dev/peps/pep-0333/#optional-platform-specific-file-handling


Download & Source
=================

The source repository is on github__.
You may submit bug reports and pull requests there.

__ https://github.com/dairiki/xsendfile_middleware/

It is also available via PyPI__.

__ https://pypi.python.org/pypi/xsendfile_middleware/


Usage
=====

The middleware is ``xsendfile_middleware.xsendfile_middleware``.
You can call it directly, e.g.::

    from xsendfile_middleware import xsendfile_middleware

    def main(**settings):
        """ Construct and return a filtered app.
        """
        def my_app(environ, start_response):
            # ...
            return app_iter

        # wrap middleware around app
        return xsendfile_middleware(my_app)

There is also a suitable entry point provided so that you can instantiate
the middleware from a PasteDeploy_ ``.ini`` file, e.g.::

    [app:myapp]
    use = egg:MyEgg

    [pipeline:main]
    pipeline = egg:xsendfile_middleware myapp

.. _PasteDeploy: http://pythonpaste.org/deploy/

Configuration
=============

X_REDIRECT_MAP
--------------

Once you have the middleware in place, the only configuration needed
(or possible) is to set an ``X_REDIRECT_MAP`` key in the WSGI environ.
How you do that depends on how you’ve got things set up.  If you are
running your app under uwsgi_, for example, then you can use something
like the following in your ``nginx`` config::

  location /app {
    uwsgi_pass 127.0.0.1:6543;
    include uwsgi_params;

    uwsgi_param X_REDIRECT_MAP /path/to/files/=/_redir_/;
  }

  location /_redir_/ {
    internal;
    alias /path/to/files/;
  }

In this configuration, if your app returns an app_iter which is
an instance of the file wrapper provided by the middleware, and
that wrapper wraps file at, *e.g.*,
``/path/to/files/dir/file.data``,
the middleware will set an ``X-Accel-Redirect: /_redir_/dir/file.data``
header in the response.  (This, hopefully, will cause nginx to send
the desired file directly to the client.)

.. _uwsgi: http://uwsgi-docs.readthedocs.org/en/latest/

The format of ``X_REDIRECT_MAP`` is a comma-separated list of
“``/prefix/=/base_uri/``” mappings.  As a short-cut, “``/prefix/``”
(with no equals sign) means the same as “``/prefix/=/prefix/``” (an
``X-Accel-Redirect`` header containing the original file name will be
sent for matching paths.)  The entries in the map are checked in
order; the first matching entry wins.  If none of the entries match,
then the middleware will pass the response on up the chain unmolested.

**Warning**:
The parsing of ``X_REDIRECT_MAP`` is rather simplistic.  Things will
probably not go well if you try to include commas, equal signs,
or any non-ASCII characters in either the *prefix* or *base_uri*.
Also, do not include any extraneous white space in ``X_REDIRECT_MAP``.

Author
======

This package was written by `Jeff Dairiki`_.

.. _Jeff Dairiki: mailto:dairiki@dairiki.org

.. |build status| image::
    https://travis-ci.org/dairiki/xsendfile_middleware.svg?branch=master
    :target: https://travis-ci.org/dairiki/xsendfile_middleware


CHANGES.rst
=======
Changes
=======

0.1.1 (2016-10-28)
==================

(Release 0.1 does not exist.  It had corrupt rST in its PKG-INFO
causing the README on PyPI not to be formatted.)

No changes (other than test configuration) from 0.1a3.  This package
has been used in production for years now — might as well drop the
“alpha” designation.

Testing
-------

* Drop support for python 3.2.  Now tested under pythons 2.6, 2.7,
  3.3, 3.4, 3.5, pypy and pypy3.

0.1a3 (2015-05-08)
==================

Brown bag.  The previous release was unusable do to this bug:

* Make sure not to include unicode strings in the headers passed to
  ``start_response``.

0.1a2 (2015-05-08)
==================

* Py3k, pypy compatibility: the tests now run under python 2.6, 2.7,
  3.2, 3.3, 3.4, pypy and pypy3.

0.1a1 (2013-12-11)
==================

Initial Release


xsendfile_middleware/test.py
# -*- coding: utf-8 -*-
"""
"""
from __future__ import absolute_import

import io
import itertools
import os
import sys
import unittest

if sys.version_info >= (3, 0):  # pragma: no cover
    unichr = chr
    text_type = str
else:
    text_type = unicode


class Test_xsendfile_middleware(unittest.TestCase):
    def make_filter(self, app):
        from xsendfile_middleware import xsendfile_middleware
        return xsendfile_middleware(app)

    def test_x_accel_redirect(self):
        def file_app(environ, start_response):
            file = DummyFile(name='/path/fn')
            start_response('200 Okay', [])
            return environ['wsgi.file_wrapper'](file, 4096)
        filtered_app = self.make_filter(file_app)
        environ = { 'X_REDIRECT_MAP': '/path/=/mapped/' }
        status, headers, app_iter = get_response(filtered_app, environ)

        redirect_uris = [value for name, value in headers
                         if name.lower() == 'x-accel-redirect']
        self.assertEqual(redirect_uris, ['/mapped/fn'])

        content_length = [value for name, value in headers
                         if name.lower() == 'content-length']
        self.assertEqual(content_length, [])

    def test_headers_are_strs(self):
        def file_app(environ, start_response):
            file = DummyFile(name='/path/fn')
            start_response('200 Okay', [])
            return environ['wsgi.file_wrapper'](file, 4096)
        filtered_app = self.make_filter(file_app)
        environ = {'X_REDIRECT_MAP': '/path/=/mapped/'}
        status, headers, app_iter = get_response(filtered_app, environ)
        for name, value in headers:
            self.assertTrue(isinstance(value, str))
            self.assertTrue(isinstance(name, str))

    def test_uses_upstream_file_wrapper_if_can_not_redirect(self):
        def file_app(environ, start_response):
            file = DummyFile(name='/path/fn')
            start_response('200 Okay', [])
            return environ['wsgi.file_wrapper'](file, 4096)
        filtered_app = self.make_filter(file_app)
        environ = {
            'wsgi.file_wrapper': DummyFileWrapper,
            'X_REDIRECT_MAP': '/does/not/match/',
            }
        status, headers, app_iter = get_response(filtered_app, environ)
        self.assertEqual(type(app_iter), DummyFileWrapper)
        self.assertEqual(app_iter.file.name, '/path/fn')

    def test_no_redirect_map(self):
        def app(environ, start_response):
            start_response('200 Okay', [])
            return iter(['body'])
        filtered_app = self.make_filter(app)
        status, headers, app_iter = get_response(filtered_app, {})
        self.assertEqual(status, '200 Okay')

    def test_non_ascii_redirect_map(self):
        def app(environ, start_response):
            start_response('200 Okay', [])
            return iter(['body'])
        filtered_app = self.make_filter(app)
        environ = {'X_REDIRECT_MAP': '/path/=/mapped/' + unichr(0xf8)}
        status, headers, app_iter = get_response(filtered_app, environ)
        self.assertEqual(status, '200 Okay')

    def test_passes_on_calls_to_write(self):
        def writer_app(environ, start_response):
            write = start_response('200 Okay', [])
            write('body')
            return iter([])
        filtered_app = self.make_filter(writer_app)
        environ = {'X_REDIRECT_MAP': '/'}
        status, headers, app_iter = get_response(filtered_app, environ)
        self.assertEqual(list(app_iter), ['body'])

    def test_start_response_called_with_exc_info(self):
        def exc_app(environ, start_response):
            try:
                raise DummyException()
            except Exception:
                start_response("200 Okay", [], sys.exc_info())
            return iter([])     # pragma: no cover
        filtered_app = self.make_filter(exc_app)
        environ = {'X_REDIRECT_MAP': '/'}
        self.assertRaises(DummyException, get_response, filtered_app, environ)

    def test_start_response_never_called(self):
        def broken_app(environ, start_response):
            return iter([])
        filtered_app = self.make_filter(broken_app)
        environ = {'X_REDIRECT_MAP': '/'}
        self.assertRaises(StartResponseNeverCalled,
                          get_response, filtered_app, environ)

    def test_start_response_called_twice(self):
        def broken_app(environ, start_response):
            start_response("200 Okay", [])
            start_response("200 Okay", [])
            return iter([])     # pragma: no cover
        filtered_app = self.make_filter(broken_app)
        environ = {'X_REDIRECT_MAP': '/'}
        self.assertRaises(StartResponseCalledTwice,
                          get_response, filtered_app, environ)

class Test_filter_app_factory(unittest.TestCase):
    def call_it(self, app, global_config, **local_conf):
        from xsendfile_middleware import filter_app_factory
        return filter_app_factory(app, global_config, **local_conf)

    def test(self):
        def app(environ, start_response):
            start_response('200 Okay', [])
            return iter([])
        filtered_app = self.call_it(app, {})
        status, headers, app_iter = get_response(filtered_app)
        self.assertEqual(status, '200 Okay')

class Test_map_filename(unittest.TestCase):
    def call_it(self, filename, redirect_map):
        from xsendfile_middleware import _map_filename
        return _map_filename(filename, redirect_map)

    def test_relative_filename(self):
        filename = 'fn'
        redirect_map = os.getcwd() + '/=/mapped/'
        uri = self.call_it(filename, redirect_map)
        self.assertEqual(uri, '/mapped/fn')

    def test_non_mapped_prefix(self):
        filename = '/path/fn'
        redirect_map = '/path/'
        uri = self.call_it(filename, redirect_map)
        self.assertEqual(uri, '/path/fn')

    def test_multiple_mappings(self):
        filename = '/path2/fn'
        redirect_map = '/path1/=/mapped1/,/path2/=/mapped2/,/path3/=/mapped3/'
        uri = self.call_it(filename, redirect_map)
        self.assertEqual(uri, '/mapped2/fn')

    def test_no_matching_mapping(self):
        filename = '/path2/fn'
        redirect_map = '/path/=/mapped/'
        uri = self.call_it(filename, redirect_map)
        self.assertEqual(uri, None)

    def test_non_ascii_filename(self):
        filename = '/' + unichr(0xf8) + '/'
        redirect_map = filename + '=/o/'
        # Convert to PEP-3333 "string of bytes"
        redirect_map = redirect_map.encode('utf8').decode('latin1')
        uri = self.call_it(filename, redirect_map)
        self.assertEqual(uri, None)

# Ripped-off from pyramid
class TestFileIter(unittest.TestCase):
    def _makeOne(self, file, block_size):
        from xsendfile_middleware import FileWrapper
        return FileWrapper(file, block_size)

    def test___iter__(self):
        f = io.BytesIO(b'abc')
        inst = self._makeOne(f, 1)
        self.assertEqual(inst.__iter__(), inst)

    def test_iteration(self):
        data = b'abcdef'
        f = io.BytesIO(b'abcdef')
        inst = self._makeOne(f, 1)
        r = b''
        for x in inst:
            self.assertEqual(len(x), 1)
            r+=x
        self.assertEqual(r, data)

    def test_close(self):
        f = io.BytesIO(b'abc')
        inst = self._makeOne(f, 1)
        inst.close()
        self.assertTrue(f.closed)

def get_response(app, environ=None):
    if environ is None:
        environ = {}

    response = []
    output = []

    def start_response(status, response_headers, exc_info=None):
        if exc_info and not output:
            reraise(*exc_info)
        elif response:
            raise StartResponseCalledTwice()

        response[:] = status, response_headers

        def save(data):
            output.append(data)
        return save

    app_iter = app(environ, start_response)
    if not response:
        raise StartResponseNeverCalled()
    status, headers = response
    if output:
        app_iter = itertools.chain(output, app_iter)
    return status, headers, app_iter

class DummyFile(object):
    def __init__(self, name):
        self.name = text_type(name)

class DummyFileWrapper(object):
    def __init__(self, file, block_size):
        self.file = file
        self.block_size = block_size

class DummyException(Exception):
    pass

class StartResponseCalledTwice(AssertionError):
    pass

class StartResponseNeverCalled(AssertionError):
    pass


def reraise(typ, value, tb):    # pragma: no cover
    # python 2/3 compatibility
    if sys.version_info < (3, 0):
        exec("raise typ, value, tb")
    else:
        assert value.__traceback__ is tb
        raise value


xsendfile_middleware/__init__.py
# -*- coding: utf-8 -*-
""" Middleware to set return x-accel-redirects for file responses.

This can be used to get nginx to send static files itself.

"""
from __future__ import absolute_import

import logging
import os

log = logging.getLogger(__name__)

def filter_app_factory(app, global_config, **local_conf):
    return xsendfile_middleware(app)

def xsendfile_middleware(application):
    def middleware(environ, start_response):
        redirect_map = environ.get('X_REDIRECT_MAP')
        if redirect_map:
            try:
                redirect_map = _ascii_str(redirect_map)
            except NotASCIIError:
                log.info("Ignoring non-ASCII value %r for X_REDIRECT_MAP",
                         redirect_map)
                redirect_map = None
        if not redirect_map:
            return application(environ, start_response)

        orig_file_wrapper = environ.get('wsgi.file_wrapper')
        environ['wsgi.file_wrapper'] = FileWrapper

        response = []
        writer = []

        def capture_response(status, response_headers, exc_info=None):
            if writer or response or exc_info:
                if response and not writer:
                    # We've been called once already, pass that call on up.
                    start_response(*response)
                writer[:] = [start_response(status, response_headers, exc_info)]
                return writer[0]        # pragma: NO COVER

            def deferred_write(data):
                if not writer:
                    writer.append(start_response(*response))
                return writer[0](data)

            response[:] = status, response_headers
            return deferred_write

        result = application(environ, capture_response)

        if writer:
            # upstream start_response already called
            return result
        elif not response:
            # Our start_response was never called. Oh well, pass the
            # problem on up.
            return result

        status, response_headers = response

        if isinstance(result, FileWrapper):
            redirect_uri = None
            name = getattr(result.file, 'name', None)
            if name:
                redirect_uri = _map_filename(name, redirect_map)
            if redirect_uri:
                # Note: Returned status seems to be ignored by nginx if
                # X-Accel-Redirect is set
                log.debug("X-Accel-Redirect to %r for %r", redirect_uri, name)
                remove_headers = ['content-length', 'x-accel-redirect']
                response_headers = [(header, value)
                                    for header, value in response_headers
                                    if header.lower() not in remove_headers]
                response_headers.append(('X-Accel-Redirect', redirect_uri))
                result = iter(())
            elif orig_file_wrapper:
                result = orig_file_wrapper(result.file, result.block_size)

        start_response(status, response_headers)
        return result

    return middleware

class NotASCIIError(Exception):
    pass

def _ascii_str(s):
    # According to PEP-3333, all "strings" used in the WSGI protocol
    # must be of type ``str``, yet they must really be a string of bytes.
    # This function converts strings to just such ``str``\s.
    #
    # As an added bit of paranoia, it ensures that the strings contain
    # nothing but ASCII characters.  (This restriction could probably
    # be relaxed, but, ATM, I'm not quite sure what encoding is
    # appropriate.)
    try:
        s = s.encode('ascii')
    except UnicodeEncodeError:
        raise NotASCIIError("%r is not an ASCII string", s)
    if not isinstance(s, str):
        # py3k
        return s.decode('latin1')  # pragma: no cover
    return s

def _map_filename(filename, redirect_map):
    filename = os.path.abspath(filename)
    try:
        filename = _ascii_str(filename)
    except NotASCIIError:
        log.info("Not mapping file with non-ASCII name %r", filename)
        return None

    for mapping in redirect_map.split(','):
        prefix, sep, base_uri = mapping.partition('=')
        if filename.startswith(prefix):
            if not sep:
                base_uri = prefix
            return base_uri + filename[len(prefix):]


_BLOCK_SIZE = 4096 * 64 # 256K

class FileWrapper(object):
    """ A fixed-block-size iterator for use as a WSGI app_iter.

    Ripped-off from pyramid.response.FileIter.

    ``file`` is a Python file pointer (or at least an object with a ``read``
    method that takes a size hint).

    ``block_size`` is an optional block size for iteration.
    """
    def __init__(self, file, block_size=_BLOCK_SIZE):
        self.file = file
        self.block_size = block_size

    def __iter__(self):
        return self

    def next(self):
        val = self.file.read(self.block_size)
        if not val:
            raise StopIteration
        return val

    __next__ = next # py3

    def close(self):
        self.file.close()
